### ISSUE TITLE
[FEATURE REQUEST] Hope to add a "secure" key import method based encryption

### ISSUE THREAD

**Issue by PIKACHUIM** [glueckkanja-pki/TPMImport#23]
## Hope to add a "secure" key negotiation based import method
First of all, please allow me to praise that this is a very good project!

Then let me talk about my request:
I have developed a TPM based smart card management tool:

https://github.com/PIKACHUIM/TPMSmartCard

This tool allows creating some virtual TPM smart cards and importing certificates

Then I encountered some trouble: _I provided a process based on C/S architecture for securely issuing certificate files to smart cards_. Users create a pair of keys locally or request the server to obtain an encrypted security key, which is then encrypted and issued by the server PFX file. 
Although PFX files use reliable and secure encryption during transmission, they always need to **be decrypted on the user's machine**, which provides an opportunity for **busers to steal PFX** (_whether they are malicious or have malicious programs on their PC_), posing a security risk to our tool.

I see that you have implemented importing certificates to TPM in another way. I would like to discuss **if there is any way to make the process of importing PFX into the system on the client safe**, such as:
1. Create a pair of RSA key pairs on TPM (or use existing ones)
2. Send the public key portion of the RSA key pair to the server
3. The server encrypts the certificate private key using the public key and then issues the key
4. After receiving the key, the client decrypts it on the TPM

This way, the private key is always encrypted to ensure that it will not be stolen
Although I have no expectations for the outcome, I am still excited to see this project and hope to receive answers or help.

I will be overwhelmed with gratitude.

**Comment by bb-froggy** [comment]
Hi!

Your TPMSmartCard project looks very sophisticated to me and seems to provide much more functionality than the simple TPMImport tool of this project. I feel honoured that you are giving it so much attention! :-)

I fully agree that with PFXes, there's always the risk that an attacker might steal the private, because while it is in the PFX, it is not secured by hardware.

It seems to me that the most direct way to mitigate this for our use case would be to create the key pair directly in the TPM, then use it to create a CSR that is sent to the CA. You pointed out this method already and the description of your TPMSmartCard project reads like it supports this approach already. Do you have a specific use case in mind where this does not suffice?

From the CA perspective, I do see these use cases (I am also a developer for the [SCEPman Cloud CA](https://www.scepman.com). For example, some clients generate CSRs that are missing extensions, Extended Key Usages or the like that are actually required for the use case. It is then sometimes easier to just create everything on the PKI server. This is the case for the YubiKey client tool, for example, where the CSRs didn't contain any extensions at all the last time I looked. But since you have a nice tool running on the client side, you could add any property needed to the CSR, so that cannot be the reason?

Apart from that, I guess what you propose is technically possible. PKCS#12 allows encrypting the content to a certificate: https://www.rfc-editor.org/rfc/rfc7292#section-3.1. I agree that this can make it more difficult for an attacker to steal the private key from the PFX. Could you explain more about the use case(s) you have in mind?

From the TPMImport side, the only change required would be to allow that when the PFX is public-key-encrypted, TPMImport could check whether a corresponding certificate is in the user/machine store and then use that to decrypt the PFX. The rest of the protocol would be out of scope for TPMImport, I believe.

**Comment by PIKACHUIM** [comment]
Thank you for your reply and support!

Yes, TPMSmartCard has implemented secure key generation and CSR on TPM smart cards, but we do have a need to issue pfx certificates and keys from the server, for example:
1. The server acts as a proxy role, not a CA, but as a certificate application proxy for users. The creation of certificates may be reviewed, created, and issued by other authoritative CAs or manually.
2. The private keys generated by TPM and CSR cannot be exported in any way, and data will be lost when users change hardware or systems. In this case, the server can generate certificates for backup and provide recovery options.

In addition, my previous question expression may have been too complicated, and in fact, it does not require many changes. Just choose one of the following two options:
1. TPMIunt or Certuti allows reading PFX information from pipelines, command lines, or TCP ports, as PFX itself is encrypted through AES and is sufficiently secure. The problem is that pfx files need to be written to disk before they can be imported, which gives any program with user privileges the opportunity to steal pfx files. (Assumption: We do not consider anyone who can bypass the operating system to read memory)
2. Create a key pair in TPMIport or Certuti, encrypt the public key to the server or the program to be imported before importing, even if the pfx file is obtained, it cannot be decrypted. This is consistent with the method you just mentioned, specific examples:
a. Create a key pair with the private key stored within the system or TPM, and print the public key on the console (as for how to interact with the server, this should be my job)
b. The server encrypts the PFX file using a public key and sends it to the client to write to the hard drive.
c. Import a file using the previously created key pair, decrypt and import it, and then delete the private key
(Additional management may need to be added here, such as which private key to use for decryption or whether only one private key can be stored)

Thank you again for your response!

**Comment by bb-froggy** [comment]
Your use case sounds absolutely valid. Some ideas:

1+TPMImport: Piping BLOBs to Win32 or .NET console applications is a bit awkward. A named pipe might be possible. Reading it from TCP ports seems like a very specific extension to a lean tool. I will think a bit more about this, but these solutions do not sound very elegant.
I have started porting TPMImport to PowerShell Core recently, but stopped because eventually it looked like I didn't need it anymore. This would make it easier to pass the PFX from memory.

1+certutil: I don't think there's a chance to accomplish this.

1 in general: As a workaround, one might create an in-memory drive, store the PFX there and import it from there. After unmounting the in-memory drive, no trace of the PFX are on the disks. Requires local admin rights on the computer, though, and drivers for the in-memory drive.

2+TPMImport:
I would modify the process you described as such:
a. Instead of creating the key pair with TPMImport, I would create a self-signed certificate in the user store with certreq or some other means (left to the user) with the private key in the TPM. You could also use an existing key if you want to renew your certificate.
b. Your job ;-)
c. When importing a PFX that turns out to be encrypted asymmetrically, TPMImport could search the appropriate store for the right key, finding the self-signed certificate/whatever was created or used in step a.. Then, it decrypts the PFX and imports the key as before.

With this approach, you can also have any number of requests in parallel, because TPMImport would just use the right key out of possibly many in the MY store. Actually, if you are renewing a key, you don't even have to do anything on the client in advance. You would just create a PFX encrypted to the existing key and deliver that to the client, where it can be decrypted securely.

2+certutil:
Windows allows encrypting a PFX to "users and groups" in the UI. I don't know how they encode that in the PFX, but it might be a way to implement your approach 2 with certutil, depending on whether a key pair is created in the backend for this.


What I don't fully get is ... you already have a beautiful client application. What do you need TPMImport for? Couldn't you achieve this completely without TPMImport?

**Comment by PIKACHUIM** [comment]
Sorry for my delayed response by a few days

1. I have conducted a brief evaluation on the method of transferring through pipelines or memory, and I believe that this approach is not very reliable or elegant. TCP is also not a good solution, and if you modify TPMIport in this way, it can only solve my problem and may not be very helpful for the TPMIport project itself. Finally, the method of creating RAM disks essentially cannot change the issue of writing files locally, and may not be suitable.

2+TPMIport: Creating certificates using Certreq or other methods and utilizing existing private keys. I think this solution may not be necessary because: 

- ① Certuti has already implemented this feature, and TPMSmartCard actually supports it.
- ② It does not seem to solve the problem of key leakage that may occur when keys are issued from the server to the local.

But I don't quite understand this sentence:

> “You would just create a PFX encrypted to the existing key and deliver that to the client, where it can be decrypted securely.”

If I have a ready-made key and certificate, it seems like I don't need to issue certificates and keys from the server anymore. Or is it necessary to create an additional certificate and key locally (as a temporary certificate) before the server issues the real certificate, and then encrypt the pfx using the public key of this temporary certificate before issuing it?

2+Certuti: means it is consistent with the 2+TPMIport scheme, but I tried to search for relevant documents, but I couldn't find them. Microsoft's documents are too difficult to find. However, when exporting PFX certificates within Windows, there is an option to use user encryption, but I can only choose local or local network users, and it seems impossible to select any remote user (not connected to the exported computer) for encryption.

Finally, I would like to explain that thank you for your praise. Although TPMIunt has a GUI, it only calls the APIs of CertUtility and OpenSC for operations. Therefore, in order to achieve end-to-end encryption and certificate issuance, other methods are needed. However, I will continue to try to study the asymmetric encryption method used by CertUtility that you mentioned

Thank you again for your reply. We welcome your communication

**Comment by PIKACHUIM** [comment]
Thank you. If possible, I would like to add TPMImport as one of the backend tools for TPMSmartCard calls. 
However, I will only add the binary file of TPMImport and the link in the project README, as TPMSmartCard uses a more "permissive" license (MIT).

I will prioritize searching for the documentation of Certutil. If Microsoft has already provided a method for remote server distribution of PFX (which I suspect may be possible as this requirement is quite common), and if it is not available, I will try to add encryption to TPMImport , but this is very difficult for me and may require a lot or most of your help.

Also, I think creating a temporary self signed certificate is a good idea, but will it increase the workload of TPMImport ? Or rather, it's simpler this way. For me, I only need to obtain a public key from the client in some way and provide it to the server when requested, so the difference between the two is not significant.

Thank you for your support, but this is not an urgent or important requirement TPMSmartCard can already meet the vast majority of requirements, regardless of whether it can be implemented in the end, I would like to express my sincere gratitude!

**Comment by PIKACHUIM** [comment]
Passing pfx data using PowerShell can indeed become very simple, but I haven't found an elegant way to pass pfx data from TPMSmartCard to PowerShell or TPMImport without writing temporary files. Do you have any good ideas? Currently, the only way I can think of is using TCP ports, which is not very elegant.

**Comment by PIKACHUIM** [comment]
Thank you very much for your proposal and work. I think it is obviously possible. Almost no certificate in DER format is larger than 4KB, and CMD can also be used. Even with a certificate link, it will not exceed 20KB. If I were to use it, I would probably use ptyprocess, so there is no length limit for CMD.

Also, I would like to know if importing certificates to **Microsoft Base Smart Card Crypto Provider** is supported. My command is:
`certutil -csp "Microsoft Base Smart Card Crypto Provider" -p "%s" -importpfx <filepath>`

Thank you again for your help!

**Comment by bb-froggy** [comment]
## Importing to Base Smart Card KSP

This should be fairly easy with an additional command line parameter that is passed in this line: https://github.com/glueckkanja-pki/TPMImport/blob/ced9de8325d125483420c2ec662a75c7c1d7b4b3/Program.cs#L151

## Code-Signing

Thank you for your offer! We also have a Code Signing certificate. If there was a demand, I could add code-signing to this project, but so far it wasn't needed. For your own uses, you can, of course, sign the executable yourself.

## Licensing -- MIT + GPL

This was in one of your previous posts actually, but I am only now realizing that your approach would violate the GPL IMHO. I am not in a position to give juridical advice, so better check for yourself, but the GPL requires that if you distribute the GPL'ed program, all programs that you distribute alongside must also be distributed using the GPL (Copyleft). This is probably not really an issue, though, as there dual-licensing should be an easy solution: You distribute TPMSmartCard, TPMImport and everything with it under the terms of the GPL; additionally, the licensee is free to choose the MIT license as an alternative for the other parts that are not TPMImport. As I said, you shouldn't rely on me for these things.

If it turns out that dual-licensing is not the easy solution I assumed it would be, please don't hesitate to come back and explain the problem. I don't want license details to get in the way of this nice use case.

**Comment by PIKACHUIM** [comment]
Thank you for your enthusiastic help. I will add TPMIport to the TPMSmartCard project when I have time.
## Licensing
I have **modified the license for the TPMSmartCard project to GPL3.0**
It is my personal project, so making this decision is easy :).

**Comment by PIKACHUIM** [comment]
However, there may be some problems:
PFX to TPM Importer
```
Unhandled exception. Internal.Cryptography.CryptoThrowHelper+WindowsCryptographicException: 
   at Internal.Cryptography.Pal.CertificatePal.FilterPFXStore(ReadOnlySpan`1 rawData, SafePasswordHandle password, PfxCertStoreFlags pfxCertStoreFlags)
   at Internal.Cryptography.Pal.CertificatePal.FromBlobOrFile(ReadOnlySpan`1 rawData, String fileName, SafePasswordHandle password, X509KeyStorageFlags keyStorageFlags)
   at System.Security.Cryptography.X509Certificates.X509Certificate..ctor(Byte[] rawData, String password, X509KeyStorageFlags keyStorageFlags)
   at System.Security.Cryptography.X509Certificates.X509Certificate2..ctor(Byte[] rawData, String password, X509KeyStorageFlags keyStorageFlags)
   at TPMImport.Program.Main(String[] args) in D:\Downloads\TPMImport-main\TPMImport-main\Program.cs:line 128

```

**Comment by PIKACHUIM** [comment]
**Ignore the previous error, my password is incorrect, this is a new error:**
```
Unhandled exception. Internal.Cryptography.CryptoThrowHelper+WindowsCryptographicException: The key set does not exist
   at System.Security.Cryptography.CngKey.Create(CngAlgorithm algorithm, String keyName, CngKeyCreationParameters creationParameters)
   at TPMImport.Program.Main(String[] args) in D:\Downloads\TPMImport-main\TPMImport-main\Program.cs:line 167
```

**Comment by bb-froggy** [comment]
Could you share your change as a Pull Request? You can mark it as "Work in Progress". This would make it easier for me to review this issue.

**Comment by PIKACHUIM** [comment]
Ok, please see this: https://github.com/glueckkanja-pki/TPMImport/pull/26
